docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Existing Solution](#existing-solution)
+    - [Deficiencies of the Existing Solution](#deficiencies-of-the-existing-solution)
+    - [Other Benefits to Replacing the Existing Solution](#other-benefits-to-replacing-the-existing-solution)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Rationale](#rationale)
+    - [Design Outline](#design-outline)
+    - [Element: New Metric in Shoot Prometheus](#element-new-metric-in-shoot-prometheus)
+    - [Element: Metric Propagation from Shoot to Seed](#element-metric-propagation-from-shoot-to-seed)
+    - [Element: Prometheus Metrics Adapter](#element-prometheus-metrics-adapter)
+    - [Element: Seed API Server Extension Service](#element-seed-api-server-extension-service)
+    - [Element: HPA](#element-hpa)
+    - [Element: VPA](#element-vpa)
+    - [Transition Strategy](#transition-strategy)
+    - [Scaling Clusters with No Monitoring Stack](#scaling-clusters-with-no-monitoring-stack)
+    - [Vision Beyond This Enhancement](#vision-beyond-this-enhancement)
+  - [Discussion and Limitations](#discussion-and-limitations)
+  - [Alternatives](#alternatives)
+  - [References](#references)
+
+## Summary
+When it comes to autoscaling shoot control plane `kube-apiserver` instances (ShootKapi hereafter), Gardener needs
+both stability and efficiency (accurate scaling). The existing approach of fusing HPA and VPA into the 2-dimensional
+autoscaler which is HVPA, poses severe algorithmic limitations which manifest as stability and efficiency issues in the
+field. 
+
+This document outlines the main deficiencies in the existing solution and proposes a simple design which overcomes them,
+reduces maintenance effort by reducing Gardener's custom code base, and enables subsequent reduction in ShootKapi
+compute cost.
+
+## Motivation
+Gardener uses a relatively complex approach for scaling shoot control plane `kube-apiserver` instances (ShootKapi
+hereafter). This need is a consequence of the following concerns:
+1. Gardener supplies K8s clusters of highly diverse compute scale and utilization - ranging from a couple of small
+   practically idle nodes, to hundreds of heavily utilized giants. Furthermore, the ratio of 'cluster compute capacity'
+   to 'ShootKapi load' substantially varies between different applications utilizing Gardener shoots. As a result,
+   compute resources required by ShootKapis of different shoots vary by two orders of magnitude.
+2. The compute cost of ShootKapi workloads constitutes a major part of Gardener's overall compute cost. Small relative
+   waste in sizing ShootKapis results in considerable absolute cost.
+3. Kube-apiserver does not scale well horizontally, in a sense that extra replicas come with substantial compute overhead.   
```

Comment:
A limited effort was made to test that with both synthetic and field tests. Also a review of related online publications. In field tests, when processing comparable request load, compared to a baseline of 1 kube-apiserver replica, 3 replicas resulted in 52-81% memory overhead, and 115-136% CPU overhead.

>Horizontal scaling always comes with overhead (base load of instances).
Certainly - the consideration is actually quantitative. The statement in the GEP is in qualitative form, strictly for brevity.

The overhead of horizontal kube-apiserver scaling was one of the most influential factors in selecting the scaling approach. Had the overhead of extra replicas been acceptable, the preferred approach would have been:
- Set vertical scale based on cluster node count, aiming at expectancy of, say, 5 replicas
- Then use only HPA to scale horizontally

Reply:
By the way, I'd be very interested to hear in case you had a different experience with compute overhead, when adding extra kube-apiserver replicas. Personally, I was surprised to see greater overhead on the CPU side (as opposed to memory). Still don't know what to make of it, but I haven't had a chance to investigate.

Reply:
Marking this discussion thread resolved to keep the PR manageable. Please don't hesitate to reopen as appropriate.